pyresample/geometry.py

```diff
@@ -717,37 +717,282 @@ def _do_transform(src, dst, lons, lats, alt):
         x, y, z = transform(src, dst, lons, lats, alt)
         return np.dstack((x, y, z))
 
-    def aggregate(self, **dims):
-        """Aggregate the current swath definition by averaging.
-
-        For example, averaging over 2x2 windows:
-        `sd.aggregate(x=2, y=2)`
+    # TODO: this is more efficient than pyproj.transform.
+    # With pyproj > 3.1 it became thread safe:
+    # https://pyproj4.github.io/pyproj/stable/advanced_examples.html#multithreading
+    # @staticmethod
+    # def swath_do_transform(src, dst, lons, lats, alt):
+    #     """Run pyproj Transformer and stack the results."""
+    #     from pyproj import Transformer
+    #     transformer = Transformer.from_crs(src.crs, dst.crs)
+    #     x, y, z = transformer.transform(lons, lats, alt, radians=False)
+    #     return np.dstack((x, y, z))
+
+    def downsample(self, x=1, y=1, **kwargs):
+        """Downsample the swath definition by averaging the coordinates along x and y dimension.
+
+        Builds upon xarray.DataArray.coarsen function.
+        To downsample of a factor of 2, call swath_def.downsample(x=2, y=2)
+        swath_def.downsample(x=1, y=1) simply returns the current swath_def.
+        By default, it raise a ValueError if the dimension size is not a multiple of the window size.
+        This can be changed by passing boundary="trim" or boundary="pad", but behaviour within pyresample is undefined.
+        See https://xarray.pydata.org/en/stable/generated/xarray.DataArray.coarsen.html for further details.
         """
         import dask.array as da
         import pyproj
+        import xarray as xr
+
+        # Check input validity
+        x = int(x)
+        y = int(y)
+        if x < 1 or y < 1:
+            raise ValueError('x and y arguments must be positive integers larger or equal to 1.')
 
+        # Return SwathDefinition if nothing to downsample
+        if x == 1 and y == 1:
+            return self
+
+        # Define geodetic and geocentric projection
         geocent = pyproj.Proj(proj='geocent')
         latlong = pyproj.Proj(proj='latlong')
+
+        # Get xr.DataArray with dask array
+        src_lats, src_lats_format = _convert_2D_array(self.lats, to='DataArray_Dask', dims=['y', 'x'])
+        src_lons, src_lons_format = _convert_2D_array(self.lons, to='DataArray_Dask', dims=['y', 'x'])
+
+        # Conversion to Geocentric Cartesian (x,y,z) CRS
```

Comment:
I'd like these operations be performed with decorators to make the code clearer. So conversion to DataArray (and back) would we one decorator, and transformation to cartesian (and back) would be the other.

Reply:
I think I see what do you mean, but not sure how to do it properly. Here I just reused the preexisting template of `Swath.Definition.aggregate `

Reply:
yes, I see that aggregate was the base for this, however now with the added array conversion it becomes clear that these conversions (array and cartesian) clutter the code and should be refactored to make the code more readable.
Using decorators, this function could be reduced to something like:
```python
@convert_arrays(...)
@convert_to_xyz(...)
def downsample(...):
    return res.coarsen(...)
```

Reply:
Mmh, this would imply that:
- **we force array and coordinate transformation also when downsample aggregation factors are 1** because AFAIK decorators would apply before the downsample function code. There is a way to avoid that? 
- we force `@convert_to_xyz(...)` to expects xr.DataArray? And raise an error if not DataArray? Or decorate  `convert_to_xyz` with `convert_arrays(to='DataArray_Dask')` decorator?

If you think this should be done @mraspaud, I can do it this afternoon. Let me know ;)